crates/libs/core/src/sync/cancel.rs

```diff
@@ -0,0 +1,729 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+use std::{
+    cell::Cell,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use mssf_com::FabricCommon::{
+    IFabricAsyncOperationCallback, IFabricAsyncOperationContext, IFabricAsyncOperationContext_Impl,
+};
+pub use tokio_util::sync::CancellationToken;
+use windows_core::{implement, AsImpl};
+
+use crate::{error::FabricErrorCode, runtime::executor::Executor};
+
+/// Async operation context for bridging rust code into SF COM api that supports cancellation.
+#[implement(IFabricAsyncOperationContext)]
+pub struct BridgeContext3<T>
+where
+    T: 'static,
+{
+    content: Cell<Option<T>>,
+    is_completed: Cell<bool>,
+    is_completed_synchronously: bool,
+    callback: IFabricAsyncOperationCallback,
+    token: CancellationToken,
+}
+
+impl<T> BridgeContext3<T>
+where
+    T: Send,
+{
+    fn new(callback: IFabricAsyncOperationCallback, token: CancellationToken) -> Self {
+        Self {
+            content: Cell::new(None),
+            is_completed: Cell::new(false),
+            is_completed_synchronously: false,
+            callback,
+            token,
+        }
+    }
+
+    /// Creates the context from callback, and returns a cancellation token that
+    /// can be used in rust code, and the cancellation token is hooked into self,
+    /// where Cancel() api cancels the operation.
+    pub fn make(callback: Option<&IFabricAsyncOperationCallback>) -> (Self, CancellationToken) {
+        let token = CancellationToken::new();
+        let ctx = Self::new(callback.unwrap().clone(), token.clone());
```

Comment:
If this code always calls ::unwrap() on `callback`, why not change the signature to `make(callback: &IFabricAsyncOperationCallback)`?

Reply:
SF com layer always uses Option<&IFabricAsyncOperationCallback> as parameter. And this api's main use case is to be used in the SF com layer. I prefer to unwrap inside here rather than defer to the caller, and the number of unwraps in code will multiply by the number of SF apis. If later we decide to check the option instead of unwrap, the this api is significantly easier to change.